src/dataflow/src/server/metrics.rs

```diff
@@ -96,183 +97,38 @@ impl WorkerMetrics {
     }
 
     fn finish(&mut self) {
-        if self.create_dataflows_int > 0 {
-            self.create_dataflows
-                .inc_by(self.create_dataflows_int as i64);
-            self.create_dataflows_int = 0;
-        }
-        if self.drop_sources_int > 0 {
-            self.drop_sources.inc_by(self.drop_sources_int as i64);
-            self.drop_sources_int = 0;
-        }
-        if self.drop_sinks_int > 0 {
-            self.drop_sinks.inc_by(self.drop_sinks_int as i64);
-            self.drop_sinks_int = 0;
-        }
-        if self.drop_indexes_int > 0 {
-            self.drop_indexes.inc_by(self.drop_indexes_int as i64);
-            self.drop_indexes_int = 0;
-        }
-        if self.peek_int > 0 {
-            self.peek.inc_by(self.peek_int as i64);
-            self.peek_int = 0;
-        }
-        if self.cancel_peek_int > 0 {
-            self.cancel_peek.inc_by(self.cancel_peek_int as i64);
-            self.cancel_peek_int = 0;
-        }
-        if self.insert_int > 0 {
-            self.insert.inc_by(self.insert_int as i64);
-            self.insert_int = 0;
-        }
-        if self.allow_compaction_int > 0 {
-            self.allow_compaction
-                .inc_by(self.allow_compaction_int as i64);
-            self.allow_compaction_int = 0;
-        }
-        if self.durability_frontier_updates_int > 0 {
-            self.durability_frontier_updates
-                .inc_by(self.durability_frontier_updates_int as i64);
-            self.durability_frontier_updates_int = 0;
-        }
-        if self.add_source_timestamping_int > 0 {
-            self.add_source_timestamping
-                .inc_by(self.add_source_timestamping_int as i64);
-            self.add_source_timestamping_int = 0;
-        }
-        if self.advance_source_timestamp_int > 0 {
-            self.advance_source_timestamp
-                .inc_by(self.advance_source_timestamp_int as i64);
-            self.advance_source_timestamp_int = 0;
-        }
-        if self.drop_source_timestamping_int > 0 {
-            self.drop_source_timestamping
-                .inc_by(self.drop_source_timestamping_int as i64);
-            self.drop_source_timestamping_int = 0;
-        }
-        if self.enable_feedback_int > 0 {
-            self.enable_feedback.inc_by(self.enable_feedback_int as i64);
-            self.enable_feedback_int = 0;
-        }
-        if self.shutdown_int > 0 {
-            self.shutdown.inc_by(self.shutdown_int as i64);
-            self.shutdown_int = 0;
-        }
-        if self.advance_all_local_inputs_int > 0 {
-            self.advance_all_local_inputs
-                .inc_by(self.advance_all_local_inputs_int as i64);
-            self.advance_all_local_inputs_int = 0;
+        for (cache_entry, counter) in self.cache.iter_mut().zip(self.counters.iter()) {
+            if *cache_entry != 0 {
+                counter.inc_by(*cache_entry);
+            }
+            *cache_entry = 0;
```

Comment:
I'm really interested in whether this does the right thing! (My assumption is it might, but will have to test) - I was surprised that mut iterator zip correctly with non-mut iterators.

Reply:
Looks like it does do the right thing! Impressive (: